### PR TITLE
[SPARK-36368][PYTHON] Fix CategoricalOps.astype to follow pandas 1.3

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR proposes to fix the behavior of `astype` for `CategoricalDtype` to follow pandas 1.3.

**Before:**
```python
>>> pcat
0    a
1    b
2    c
dtype: category
Categories (3, object): ['a', 'b', 'c']

>>> pcat.astype(CategoricalDtype(["b", "c", "a"]))
0    a
1    b
2    c
dtype: category
Categories (3, object): ['b', 'c', 'a']
```

**After:**
```python
>>> pcat
0    a
1    b
2    c
dtype: category
Categories (3, object): ['a', 'b', 'c']

>>> pcat.astype(CategoricalDtype(["b", "c", "a"]))
0    a
1    b
2    c
dtype: category
Categories (3, object): ['a', 'b', 'c']  # CategoricalDtype is not updated if dtype is the same
```

`CategoricalDtype` is treated as a same `dtype` if the unique values are the same.

```python
>>> pcat1 = pser.astype(CategoricalDtype(["b", "c", "a"]))
>>> pcat2 = pser.astype(CategoricalDtype(["a", "b", "c"]))
>>> pcat1.dtype == pcat2.dtype
True
```

### Why are the changes needed?

We should follow the latest pandas as much as possible.

### Does this PR introduce _any_ user-facing change?

Yes, the behavior is changed as example in the PR description.

### How was this patch tested?

Unittest

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @xinrong-databricks @ueshin FYI

**Review by HyukjinKwon — APPROVED** [review]

**Comment by itholic** [comment]
I'm fixing CategoricalIndexTest.

File: python/pyspark/pandas/tests/test_categorical.py [hunk 1]
```diff
@@ -243,18 +243,16 @@ def test_astype(self):
...
-            # TODO(SPARK-36367): Fix the behavior to follow pandas >= 1.3
-            pass
-        elif LooseVersion(pd.__version__) >= LooseVersion("1.2"):
             self.assert_eq(
                 kcser.astype(CategoricalDtype(["b", "c", "a"])),
                 pcser.astype(CategoricalDtype(["b", "c", "a"])),
             )
         else:
             self.assert_eq(
                 kcser.astype(CategoricalDtype(["b", "c", "a"])),
```
Comment (xinrong-databricks) [inline python/pyspark/pandas/tests/test_categorical.py:254]: Shall we rename `kcser` to `pscser`? It's fine to do it in a separate PR though.

File: python/pyspark/pandas/tests/test_categorical.py [hunk 2]
```diff
@@ -243,18 +243,16 @@ def test_astype(self):
...
-                pser.astype(CategoricalDtype(["b", "c", "a"])),
+                kcser,
```
Comment (xinrong-databricks) [inline python/pyspark/pandas/tests/test_categorical.py:255]: So when `LooseVersion(pd.__version__) < LooseVersion("1.3")`, `kcser` is not updated?

Comment (ueshin) [reply python/pyspark/pandas/tests/test_categorical.py:255]: btw, nit: this should be `pcser` instead of `kcser`?

**Comment by ueshin** [comment]
We should also fix tests in `python/pyspark/pandas/tests/indexes/test_category.py` and `python/pyspark/pandas/tests/data_type_ops/test_categorical_ops.py` as well?

File: python/pyspark/pandas/tests/test_categorical.py (hunk 1 above)
Comment (itholic) [reply python/pyspark/pandas/tests/test_categorical.py:254]: Sure!

File: python/pyspark/pandas/tests/test_categorical.py (hunk 2 above)
Comment (itholic) [reply python/pyspark/pandas/tests/test_categorical.py:255]: Yes, because `kcser` is pandas-on-Spark Series, which always follows latest pandas(1.3).

**Comment by itholic** [comment]
> We should also fix tests in `python/pyspark/pandas/tests/indexes/test_category.py` and `python/pyspark/pandas/tests/data_type_ops/test_categorical_ops.py` as well?

Definitely! Let me address them as well.

File: python/pyspark/pandas/tests/data_type_ops/test_categorical_ops.py [hunk 3]
```diff
@@ -183,22 +183,31 @@ def test_astype(self):
...
+            self.assert_eq(psser.astype(np.int32), ps.Series(data, dtype=np.int32))
+            self.assert_eq(psser.astype(np.int16), ps.Series(data, dtype=np.int16))
+            self.assert_eq(psser.astype(np.int8), ps.Series(data, dtype=np.int8))
+            self.assert_eq(psser.astype(bool), ps.Series(data, dtype=bool))
+        else:
+            self.assert_eq(pser.astype(np.float32), psser.astype(np.float32))
+            self.assert_eq(pser.astype(np.int32), psser.astype(np.int32))
+            self.assert_eq(pser.astype(np.int16), psser.astype(np.int16))
+            self.assert_eq(pser.astype(np.int8), psser.astype(np.int8))
+            self.assert_eq(pser.astype(bool), psser.astype(bool))
```
Comment (ueshin) [inline python/pyspark/pandas/tests/data_type_ops/test_categorical_ops.py:203]: Do we need this? It seems working with pandas 1.2 without this change.

File: python/pyspark/pandas/tests/data_type_ops/test_categorical_ops.py [hunk 4]
```diff
@@ -183,22 +183,31 @@ def test_astype(self):
...
         cat_type = CategoricalDtype(categories=[3, 1, 2])
+        # CategoricalDtype is not updated if the dtype is same from pandas 1.3.
         if LooseVersion(pd.__version__) >= LooseVersion("1.3"):
-            # TODO(SPARK-36367): Fix the behavior to follow pandas >= 1.3
-            pass
-        elif LooseVersion(pd.__version__) >= LooseVersion("1.2"):
             self.assert_eq(pser.astype(cat_type), psser.astype(cat_type))
         else:
-            self.assert_eq(pd.Series(data).astype(cat_type), psser.astype(cat_type))
+            self.assert_eq(psser.astype(cat_type), psser)
```
Comment (ueshin) [inline python/pyspark/pandas/tests/data_type_ops/test_categorical_ops.py:210]: nit: the second argument should be `pser` instead of `psser`?

File: python/pyspark/pandas/tests/test_categorical.py [hunk 5]
```diff
@@ -239,25 +239,23 @@ def test_astype(self):
...
-                kcser.astype(CategoricalDtype(["b", "c", "a"])),
+                pscser.astype(CategoricalDtype(["b", "c", "a"])),
                 pcser.astype(CategoricalDtype(["b", "c", "a"])),
             )
         else:
             self.assert_eq(
-                kcser.astype(CategoricalDtype(["b", "c", "a"])),
-                pser.astype(CategoricalDtype(["b", "c", "a"])),
+                pscser.astype(CategoricalDtype(["b", "c", "a"])),
+                pscser,
```
Comment (ueshin) [inline python/pyspark/pandas/tests/test_categorical.py:255]: This should be `pcser`?

File: python/pyspark/pandas/tests/indexes/test_category.py [hunk 6]
```diff
@@ -176,18 +176,16 @@ def test_astype(self):
...
-            # TODO(SPARK-36367): Fix the behavior to follow pandas >= 1.3
-            pass
-        elif LooseVersion(pd.__version__) >= LooseVersion("1.2"):
             self.assert_eq(
                 kcidx.astype(CategoricalDtype(["b", "c", "a"])),
                 pcidx.astype(CategoricalDtype(["b", "c", "a"])),
             )
         else:
             self.assert_eq(
                 kcidx.astype(CategoricalDtype(["b", "c", "a"])),
```
Comment (ueshin) [inline python/pyspark/pandas/tests/indexes/test_category.py:187]: Shall we rename this `kcidx` as well?

File: python/pyspark/pandas/tests/indexes/test_category.py [hunk 7]
```diff
@@ -176,18 +176,16 @@ def test_astype(self):
...
-                pidx.astype(CategoricalDtype(["b", "c", "a"])),
+                kcidx,
```
Comment (ueshin) [inline python/pyspark/pandas/tests/indexes/test_category.py:188]: This should be `pcidx`?

File: python/pyspark/pandas/tests/data_type_ops/test_categorical_ops.py (hunk 3 above)
Comment (itholic) [reply python/pyspark/pandas/tests/data_type_ops/test_categorical_ops.py:203]: Oh... then seems like we're missing tests for pandas 1.2 ??

It's failed on my local with pandas 1.2.

```
Attribute "dtype" are different
[left]:  float64
[right]: float32

Left:
0    1.0
1    2.0
2    3.0
dtype: float64
float64

Right:
0    1.0
1    2.0
2    3.0
dtype: float32
float32
```

The result is actually different.

```python
>>> pd.__version__
'1.2.0'
>>> pser.astype(np.float32)
0    1.0
1    2.0
2    3.0
dtype: float64
>>> psser.astype(np.float32)
0    1.0
1    2.0
2    3.0
dtype: float32
```

Comment (ueshin) [reply python/pyspark/pandas/tests/data_type_ops/test_categorical_ops.py:203]: Seems like it's a pandas' bug?
`pser.astype(np.float32)` should definitely return `dtype: float32`..
Could you narrow down the version range then?

Comment (ueshin) [reply python/pyspark/pandas/tests/data_type_ops/test_categorical_ops.py:203]: pandas 1.2.5 seems working, btw.

Comment (itholic) [reply python/pyspark/pandas/tests/data_type_ops/test_categorical_ops.py:203]: @ueshin Oh, yes I'll narrow down the version range after checking the all 1.2.x versions.

Thanks!

File: python/pyspark/pandas/tests/indexes/test_category.py [hunk 8]
```diff
@@ -176,18 +176,16 @@ def test_astype(self):
 
         self.assert_eq(kcidx.astype("category"), pcidx.astype("category"))
 
+        # CategoricalDtype is not updated if the dtype is same from pandas 1.3.
         if LooseVersion(pd.__version__) >= LooseVersion("1.3"):
-            # TODO(SPARK-36367): Fix the behavior to follow pandas >= 1.3
-            pass
-        elif LooseVersion(pd.__version__) >= LooseVersion("1.2"):
             self.assert_eq(
                 kcidx.astype(CategoricalDtype(["b", "c", "a"])),
```
Comment (ueshin) [inline python/pyspark/pandas/tests/indexes/test_category.py:182]: We should rename this and the definition to `pscidx` too?

Comment (itholic) [reply python/pyspark/pandas/tests/indexes/test_category.py]: yeah, just fixed it. Thanks!

**Review by ueshin — APPROVED** [review]
LGTM, pending tests.

File: python/pyspark/pandas/tests/test_categorical.py (hunk 2 above)
Comment (xinrong-databricks) [reply python/pyspark/pandas/tests/test_categorical.py]: Oh makes sense, thanks for the explanation.

**Review by xinrong-meng — APPROVED** [review]

**Comment by ueshin** [comment]
Thanks! merging to master.
